sympy/combinatorics/homomorphisms.py

```diff
@@ -417,3 +419,88 @@ def block_homomorphism(group, blocks):
     images = {g: Permutation([b[p[i]^g] for i in identity]) for g in group.generators}
     H = GroupHomomorphism(group, codomain, images)
     return H
+
+def group_isomorphism(G, H):
+    '''
+    Compute an isomorphism between 2 given groups.
+
+    Arguments:
+        G (a finite `FpGroup` or a `PermutationGroup`) -- First group
+        H (a finite `FpGroup` or a `PermutationGroup`) -- Second group
+
+    Returns:
+    If the groups are not isomorphic, `None` is returned.
+    If the groups are isomorphic, the corresponding `GroupHomomorphism` object is returned.
+
+    Summary:
+    Uses the approach suggested by Robert Tarjan to compute the isomorphism between two groups.
+    First, the set of generators are mapped with the elements of `H` and we check if the mapping induces an isomorphism.
+
+    '''
+
+    if isinstance(G, FpGroup):
+        if G.order() == S.Infinity:
+            raise NotImplementedError("Isomorphism methods are not implemented for infinite groups.")
+        G = G._to_perm_group()[0]
+    if isinstance(H, FpGroup):
+        if H.order() == S.Infinity:
+            raise NotImplementedError("Isomorphism methods are not implemented for infinite groups.")
+        H = H._to_perm_group()[0]
+
+    if not isinstance(G, PermutationGroup):
+        raise TypeError("The group must be a PermutationGroup or a finite FpGroup")
+    if not isinstance(H, PermutationGroup):
+        raise TypeError("The group must be a PermutationGroup or a finite FpGroup")
+
+    if (len(G) != len(H)) and (G.is_abelian != H.is_abelian):
+        return None
+
+    # Match the generators of `G` with the subsets of H
+    gens = G.generators
+    for subset in itertools.permutations(H, len(gens)):
+        # Create a map
+        func_map = dict(zip(gens, subset))
+        counterexample = False
+        # Loop through the mapped elements and try to extend the mapping
+        # or to find a counter example.
+        while not counterexample:
+            extended_map = {}
+            for g, h in itertools.product(func_map, func_map):
+                if g*h in func_map:
+                    if func_map[g]*func_map[h] != func_map[g*h]:
+                        counterexample = True
+                        break
+                else:
+                    extended_map[g*h] = func_map[g]*func_map[h]
+
+            # Break when all the elements are mapped.
+            if len(func_map) == len(G):
+                break
+
+            # Remove the duplicate elements in extended_map
+            # and merge the extended_map with the map.
+            image_len = len(set(func_map.values()) + set(extended_map.values()))
+            if image_len != len(func_map) + len(extended_map):
+                counterexample = True
+                break
+            func_map.update(extended_map)
```

Comment:
Should you ju-st break here? Presumably `func_map` defines an isomorphism if it got to this stage

Reply:
Ah, no you shouldn't, because you are waiting for it to become surjective first.